src/Servers/HttpSys/src/HttpSysListener.cs

```diff
@@ -310,103 +309,79 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+                bytes = allocator.GetHeaderEncodedBytes(headerValue, out bytesLength);
+                nativeHeaderValues[header->KnownHeaderCount].RawValueLength = checked((ushort)bytesLength);
+                nativeHeaderValues[header->KnownHeaderCount].pRawValue = bytes;
+                header->KnownHeaderCount++;
             }
 
-            httpResponse.Response_V1.StatusCode = (ushort)httpStatusCode;
-            string? statusDescription = HttpReasonPhrase.Get(httpStatusCode);
-            uint dataWritten = 0;
-            uint statusCode;
-            byte[] byteReason = statusDescription != null ? HeaderEncoding.GetBytes(statusDescription) : Array.Empty<byte>();
-            fixed (byte* pReason = byteReason)
-            {
-                httpResponse.Response_V1.pReason = (byte*)pReason;
-                httpResponse.Response_V1.ReasonLength = (ushort)byteReason.Length;
+            knownHeaderInfo[0].pInfo = header;
 
-                byte[] byteContentLength = new byte[] { (byte)'0' };
-                fixed (byte* pContentLength = byteContentLength)
-                {
-                    (&httpResponse.Response_V1.Headers.KnownHeaders)[(int)HttpSysResponseHeader.ContentLength].pRawValue = (byte*)pContentLength;
-                    (&httpResponse.Response_V1.Headers.KnownHeaders)[(int)HttpSysResponseHeader.ContentLength].RawValueLength = (ushort)byteContentLength.Length;
-                    httpResponse.Response_V1.Headers.UnknownHeaderCount = 0;
-
-                    statusCode =
-                        HttpApi.HttpSendHttpResponse(
-                            _requestQueue.Handle,
-                            requestId,
-                            0,
-                            &httpResponse,
-                            null,
-                            &dataWritten,
-                            IntPtr.Zero,
-                            0,
-                            SafeNativeOverlapped.Zero,
-                            IntPtr.Zero);
-                }
-            }
-            if (statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS)
-            {
-                // if we fail to send a 401 something's seriously wrong, abort the request
-                HttpApi.HttpCancelHttpRequest(_requestQueue.Handle, requestId, IntPtr.Zero);
-            }
+            httpResponse.ResponseInfoCount = 1;
         }
-        finally
+
+        httpResponse.Response_V1.StatusCode = checked((ushort)httpStatusCode);
+        string statusDescription = HttpReasonPhrase.Get(httpStatusCode) ?? string.Empty;
+        uint dataWritten = 0;
+        uint statusCode;
+
+        bytes = allocator.GetHeaderEncodedBytes(statusDescription, out bytesLength);
+        httpResponse.Response_V1.pReason = bytes;
+        httpResponse.Response_V1.ReasonLength = checked((ushort)bytesLength);
+
+        int pContentLengthLength = 1;
+        byte* pContentLength = stackalloc byte[pContentLengthLength + 1];
```

Comment:
stackalloc is more efficient with constants right?
```suggestion
        byte* pContentLength = stackalloc byte[2];
```

Reply:
I've marked `pContentLengthLength` as `const` and with the `+ 1` this should be obvious to the JIT now.

Reply:
I expect that the constant-prop optimization will be done in Roslyn when this is marked const. Constant-prop is one of the few optimizations done by Roslyn.

Rename `pContentLengthLength` to `contentLengthLength`? The variable is not a pointer, so it should not have `p` prefix.